shaping/wrapping.go

```diff
@@ -557,36 +557,77 @@ const (
 	newCandidate
 )
 
-// fillUntil tries to fill the provided line candidate slice with runs until it reaches a run containing the
-// provided break option. It returns the index of the run containing the option, the new width of the candidate
-// line, the contents of the new candidate line, and a result indicating how to proceed.
-func (l *LineWrapper) fillUntil(runs RunIterator, option breakOption, startWidth fixed.Int26_6, lineCandidate []Output) (newWidth fixed.Int26_6, newLineCandidate []Output, status fillResult) {
+// lineConfig tracks settings for line wrapping a single line of text.
+type lineConfig struct {
+	// truncating indicates whether this line is being truncated (if sufficiently long).
+	truncating bool
+	// maxWidth is the maximum space a line can occupy.
+	maxWidth int
+	// truncatedMaxWidth holds the maximum width of the line available for text if the truncator
+	// is occupying part of the line.
+	truncatedMaxWidth int
+}
+
+// lineBuilder manages line wrapping candidates.
+type lineBuilder struct {
+	// best tracks the best known line built by this builder within the maxWidth it was used with.
+	best []Output
+	// candidate holds intermediate line state as new lines are built and tested to fit.
+	candidate []Output
+	// candidateWidth tracks the total advance of the runs in candidate.
+	candidateWidth fixed.Int26_6
+}
+
+// line returns the best known accumulated line in the builder.
+func (l *lineBuilder) line() []Output {
+	if len(l.best) > 0 {
+		return l.best
+	}
+	return l.candidate
+}
+
+// appendCandidate updates l.candidate to add run to the end and updates l.candidateWidth
+// with the new width of the line candidate.
+func (l *lineBuilder) appendCandidate(run Output) {
+	l.candidate = append(l.candidate, run)
+	l.candidateWidth += run.Advance
+}
+
+// commitCandidate updates l.best to be the results of concatenating withNewRun to
+// the end of l.candidate, but does not modify l.candidate. This allows us to save
+// a known good line while also perserving our ability to try other suffixes on the
+// end of l.candidate.
+func (l *lineBuilder) commitCandidate(withNewRun Output) {
+	l.best = commitCandidate(l.best, l.candidate, withNewRun)
```

Comment:
Maybe merge `commitCandidate` into this method ? Unless you expect to use it later ?
I also note that `commitCandidate` is only used with a single `Output` (again, for now) so you could simplify its signature and implementation.

Reply:
True, I think it would probably make sense to move that here. Thanks!